### ISSUE TITLE
[Documentation Update Request] Schema index section needs to be updated

### ISSUE THREAD

**Issue by samlaf** [dynamoose/dynamoose#1239]
The [index section](https://dynamoosejs.com/guide/Schema#index-boolean--object--array) of the schema documentation I believe is not true.
Everything in there mentions the index being created.
I was scared that if I added the index attribute it would recreate the index everytime I try to use dynamoose... however this doesn't seem to be the case. If the index is already present, it seems so far like dynamoose will just use that index.

It would be nice to update the documentation so as not to scare people away from using this feature!
Thank you. :)

**Comment by fishcharlie** [comment]
@samlaf 

> Everything in there mentions the index being created

In the first line of that documentation it states:

> You can define indexes on properties to be created or updated upon model initialization.

Just curious, how would you improve this? To me this seems to indicate that it won't be recreated every time due to the `updated` language. Just trying to figure out the best way to add additional clarity here.

**Comment by samlaf** [comment]
I might be totally misunderstanding what dynamoose is doing here.
I thought the 3 choices would (should) be: create, update, or just use.

Suppose I already have an index called byAge on my dynamodb table.
If I don't add an index object when building my dynamoose schema, getIndexes() returns an empty list.
If I do add the index object however, what will this do to my dynamodb table? Will it just read the index and make sure it is there, and then be able to use it, or does it ABSOLUTELY have to create a new one or update the existing one (what if there is nothing to update...?)

**Comment by fishcharlie** [comment]
@samlaf That is correct. The purpose of the index is to match what you have in your schema to the DynamoDB table.

To me, if it matches already, `update` indicates that it won't take any action.

Do you believe that changing that line to state the following is more clear?

> You can define indexes on properties to be created or updated (if necessary) upon model initialization.

**Comment by samlaf** [comment]
Why does 'update' sound like it doesn't take any action to you? It sounds
to me like it's going to update something... which it won't.

Let me break down my understanding of what dynamoose is doing, to make sure
we're on the same page and I'm not misunderstanding something.
There are 2 entities that we need to talk about:

   1. the actual dynamodb table stored on aws servers
   2. the in-memory ORM object that dynamoose creates to simplify
   communicating with (1)

(2) enforces a schema on top of (1). Suppose we want to create the
in-memory object (2) for a dynamodb table called "Cat" which has a simple
primary key consisting of hash key "name" (no sort key), and a global
secondary index on "color" (no hash key) named "byColor".

const schema = {"name": String, "color": String};

const catModel = dynamoose.model("Cat", schema);


 When doing so, it is possible for the dynamoose catModel (1) to
communicate with the Cat dynamodb table (2) and ask to list its index.
Hence, I was expecting for catModel.getIndexes() to return "byColor".
However, it doesn't, and we need to change the schema to look something like

> const schema = {"name": String, "color": {"type": String, "index":
> {"global": true, "name": "byColor"}}};

and then our catModel will have access to (1)'s GSI. But does using
catModel with this schema "update" the index on (1)? Absolutely not!

 Here are the 3 use cases that I see:
1. create

> const schema = {"name": String, "color": String, "age": {"type": Number,
> "index": {"global": true, "name": "byAge"}}};

This would create the new index "byAge" since it doesn't exist on (1)

2. update

>  const schema = {"name": String, "color": {"type": String, "index":
> {"global": false, "name": "byColor"}}};

This would update the "byColor" index to be local instead of global (note
that this ACTUALLY changes the dynamodb model, hence "update")

3. locate

>  const schema = {"name": String, "color": {"type": String, "index":
> {"global": true, "name": "byColor"}}};

Here the index "byColor" already exists on (2), and we are just loading it
in memory (1) to be able to reference it and make queries on it.




Hence, based on this, I might propose changing the documentation to
something like:

> Indexes defined on your DynamoDB tables must be mentionned in your
> dynamoose schema; dynamoose does not automatically read them. In the case
> that a schema index does not already exist on the DynamoDB table, it will
> be created on first use (or model initialization?). If it already exists
> but is different, the DynamoDB index will be updated on first use.
>

**Comment by fishcharlie** [comment]
@samlaf 

> Why does 'update' sound like it doesn't take any action to you?

Not sure I can answer this. Maybe because I wrote both the code and documentation so I'm just aware of how it works? Which is why I'm glad you submitted this issue.

> Let me break down my understanding of what dynamoose is doing

I think the easiest way to think about this is that Dynamoose will update the DynamoDB indexes to match whatever you set in Dynamoose.

So yes, I believe you are correct. Remember tho, DynamoDB does not have a concept of "updating an index". So it'd really be deleting indexes that exist on the table but don't exist in Dynamoose, and creating indexes that exist in Dynamoose and don't exist on DynamoDB.

Now, one edge case. If you have the `update` and `create` model options set to false, it won't take any action in regards to indexes.

---

Anyways. I think we'll update it to the following:

>  Indexes on your DynamoDB tables must be defined in your Dynamoose schema. If you have the `update` option set to `true` on your model settings, and a Dynamoose schema index does not already exist on the DynamoDB table, it will be created on model initialization. Similarily, indexes on your DynamoDB table that do not exist in your Dynamoose schema will be deleted.

We will then remove the following line:

> You can define indexes on properties to be created or updated upon model initialization.

And the rest of that first paragraph will be separated into a new paragraph (the second paragraph). 

If you would like to submit a PR for this, feel free. Otherwise I'll get around to it when I can.

**Comment by samlaf** [comment]
I was going to get back to you but it totally slipped my mind...
since I'm again running out of time right now, I'll just say that I like
this update! And really terribly sorry for 'schooling' you about your own
project (I should have digged deeper into who was answering this issue haha)
I'll let you do the PR if you don't mind... thank you for the amazing work
with this library!

**Comment by fishcharlie** [comment]
@samlaf 

> I'll just say that I like this update
> I'll let you do the PR if you don't mind

No problem. I'll try to make time to do this PR soon.

> And really terribly sorry for 'schooling' you about your own project

Technically I didn't write this project originally, but am the main maintainer these days 😃.

> thank you for the amazing work with this library

Of course! Keep the feedback coming. Really appreciate it.
